packages/@lwc/compiler/src/transformers/template.ts

```diff
@@ -70,18 +70,13 @@ function serialize(
         `${namespace}-${name}_${path.basename(filename, path.extname(filename))}`
     );
     let buffer = '';
-    buffer += `import _implicitStylesheets from "${cssRelPath}";\n\n`;
+    buffer += `import { registerStylesheets } from "lwc";\n`;
+    buffer += `import _implicitStylesheets from "${cssRelPath}";\n`;
     buffer += `import _implicitScopedStylesheets from "${scopedCssRelPath}?scoped=true";\n\n`;
     buffer += code;
     buffer += '\n\n';
-    buffer += 'if (_implicitStylesheets) {\n';
-    buffer += `  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)\n`;
-    buffer += `}\n`;
-    buffer += 'if (_implicitScopedStylesheets) {\n';
-    buffer += `  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitScopedStylesheets)\n`;
-    buffer += `}\n`;
     buffer += 'if (_implicitStylesheets || _implicitScopedStylesheets) {\n';
-    buffer += `  tmpl.stylesheetToken = "${scopeToken}"\n`;
+    buffer += `  registerStylesheets(tmpl, "${scopeToken}", _implicitStylesheets, _implicitScopedStylesheets);\n`;
```

Comment:
Instead of exposing this new method called `registerStylesheets`, what do you think about moving the generation of the stylesheet imports into the template compiler and extends the `registerTemplate` method to handle this?

```js
registerTemplate( tmp, {
  slot: [],
  token: "some token",
  stylesheets: [/* imported stylesheets */]
})
```

It would require passing new information to the template compiler like the filename and some scoping token to make this possible.

Reply:
Yep I don't see why we couldn't do that. It would make sense not to have an extra exported API.

Reply:
Hm, actually it looks like the `registerTemplate` call comes from `@lwc/template-compiler`, but the `tmpl.stylesheets = ...` comes from `@lwc/compiler`. There may be a way to refactor this to work, though. I'll look into it.

Reply:
After looking at this again, it looks technically feasible but difficult.

The stylesheets and token are attached to the template here:

https://github.com/salesforce/lwc/blob/b41c7dd8e7621c7e799d4ffe0e1b99fbd6a1b14a/packages/%40lwc/compiler/src/transformers/template.ts#L73-L85

And the `registerTemplate` is generated here:

https://github.com/salesforce/lwc/blob/dbf80c5e769dc663c9c47e396a6447d183613876/packages/%40lwc/template-compiler/src/codegen/formatters/module.ts#L70-L74

Note that the `@lwc/template-compiler` exports a template that is already "secure" (via `registerTemplate`).

To change this, we would need to either 1) have the `@lwc/template-compiler` import the stylesheets, or 2) export an insecure template from `@lwc/template-compiler` and have `@lwc/compiler` call `registerTemplate`.

Either one is technically feasible, but I worry that it would interact poorly with the ecosystem built around both packages. I suppose we can try it out and see if Nucleus breaks.

My vote would be for # 1 since it would be far less refactoring, and I don't think anyone is using `@lwc/template-compiler` directly and expecting it to be secure (I hope?).

Reply:
Sounds good to me.

Reply:
I played around with this, but now I'm not sure that either option 1 or 2 make much sense. If we move setting the `stylesheetToken`/`stylesheet` into `registerTemplate`, then the ACT compiler will always generate an error message because it sets `tmpl.stylesheets = ...` after calling `registerTemplate`. So we would potentially have to update it. (Or deal with lots of error messages in our ACT Karma tests.)

Plus, we do not save much by moving `tmpl.slots = [...]` and `tmpl.renderMode = 'light'` into `registerTemplate`. These don't take up much space because they only apply to templates with slots or in light mode.

Conceptually I like the idea of having `registerTemplate()` do all this `tmpl` setup, but it might be really tricky. I'll keep poking at it though.

Reply:
Here is the commit to move everything to `registerTemplate`: ba15ba0a0

Another downside is that we now have to add new required options to the template compiler API: `stylesheetToken` and `filename`. (Potentially we could just add `name` and `namespace` instead of `stylesheetToken`, but we still need to add something so that it can generate the stylesheetToken and stylesheet imports.)

These two options (`registerStylesheets` vs move everything to `registerTemplate`) have pros and cons. I'm not sure which one is better, but I'm leaning toward `registerStylesheets`.